[ALLUXIO-1871] Improve Alluxio scripts for non-sudo deployment

peisun1115 opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-1871
Right now sudo access is required to start Alluxio because of ramfs. This issue is to track tasks to make it easier to start alluxio for people without sudo access.

Summary:
1.  Use tmmFS (pre-mounted on /dev/shm) in linux machines. 
2. Mac OS-X is not supported.

How to use: (Linux only)
./bin/alluxio-start.sh local NoMount

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9202/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9203/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9206/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9206/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9206/org.alluxio$alluxio-tests/testReport/alluxio.client.concurrent/FileInStreamConcurrencyIntegrationTest/FileInStreamConcurrencyTest/'><strong>alluxio.client.concurrent.FileInStreamConcurrencyIntegrationTest.FileInStreamConcurrencyTest</strong></a></li></ul>
Test FAILed.

gpang commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9211/
Test PASSed.

gpang commented on bin/alluxio-start.sh:
```diff
@@ -12,18 +12,19 @@ BIN=$(cd "$( dirname "$0" )"; pwd)
-  workers MOPT\t\tStart workers on worker nodes
-  restart_worker\tRestart a failed worker on this node
-  restart_workers\tRestart any failed workers on worker nodes
+  local [MOPT] \t\t\tStart a master and worker locally. SudoMount is assumed is MOPT is missing.
```
"is MOPT" -> "if MOPT"

peisun1986 replied:
Done

gpang commented on bin/alluxio-start.sh:
```diff
@@ -12,18 +12,19 @@ BIN=$(cd "$( dirname "$0" )"; pwd)
   Mount\t\t\tMount the configured RamFS. Notice: this will format the existing RamFS.
   SudoMount\t\tMount the configured RamFS using sudo. Notice: this will format the existing RamFS.
-  NoMount\t\tDo not mount the configured RamFS
+  NoMount\t\tDo not mount the configured RamFS. Notice: Use NoMount and set ALLUXIO_RAM_FOLDER to
```
I have a high-level question. Is it possible to make this work such that if NoMount is used, it will automatically use tmpFS instead of requiring the user to set the appropriate env variable? That could make this very user friendly.

peisun1986 replied:
Yes. Done.

gpang replied:
I want to clarify. In your description, you said the user would have to use:

ALLUXIO_RAM_FOLDER=/dev/shm ./bin/alluxio-start.sh all NoMount

However, it is also possible to leave out ALLUXIO_RAM_FOLDER, like: 

./bin/alluxio-start.sh all NoMount

and will it still work? I think this is where we want to ultimately get to. For example, if we run:

./bin/alluxio-start.sh local

if it is on OSX, it will use the ramdisk at '/Volumes/ramdisk'
If it is on linux, it uses the ramdisk at 'mnt/ramdisk'
This is transparent to the user.

Does this PR allow this transparency with NoMount?

peisun1986 replied:
Yes, it works. I forgot to remove ALLUXIO_RAM_FOLDER=/dev/shm from the PR description.  Updated.

gpang commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,20 @@ check_mount_mode() {
+          echo -e "${USAGE}"
+          exit 1
+        fi
+        echo "Warning: Alluxio is running on tmpfs that supports swapping."
```
add something like "is not guaranteed to be in memory"

peisun1986 replied:
Done

gpang commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,20 @@ check_mount_mode() {
+          exit 1
+        fi
+        echo "Warning: Alluxio is running on tmpfs that supports swapping."
+        echo "Warning: Check vmstat if Alluxio is slow."
```
maybe, "Check vmstat for memory statistics.

peisun1986 replied:
Done

gpang commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,20 @@ check_mount_mode() {
+        echo "Warning: Alluxio is running on tmpfs that supports swapping."
+        echo "Warning: Check vmstat if Alluxio is slow."
+      else
+        echo "WARNING: using NoMount but ALLUXIO_RAM_FOLDER is not set to /dev/shm."
```
Could ALLUXIO_RAM_FOLDER be set to the ramdisk folder?

peisun1986 replied:
Are you asking whether  ALLUXIO_RAM_FOLDER can be set to /mnt/ramdisk here? I am not sure I understand you question.

gpang replied:
I was just asking that it is probably possible if the user executed:

ALLUXIO_RAM_FOLDER=/mnt/ramdisk ./bin/alluxio-start.sh all NoMount

Then, does the warning make sense?

peisun1986 replied:
I have removed the warning to avoid confusion.

ALLUXIO_RAM_FOLDER=/mnt/ramdisk ./bin/alluxio-start.sh all NoMount is not allowed unless ramFS was mounted manually by the user.

gpang commented:
@peisun1986 Thanks for the PR! I had a high-level question about the possibility of making this even more user friendly.

peisun1986 commented:
Gene, thanks for the comments. I have addressed them. PTAL?

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9213/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9213/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/9213/org.alluxio$alluxio-tests/testReport/alluxio.client.concurrent/FileInStreamConcurrencyIntegrationTest/FileInStreamConcurrencyTest/'><strong>alluxio.client.concurrent.FileInStreamConcurrencyIntegrationTest.FileInStreamConcurrencyTest</strong></a></li></ul>
Test FAILed.

gpang commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9215/
Test PASSed.

gpang commented:
@peisun1986 I just had a clarifying question about one of your comments. Thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9217/
Test PASSed.

peisun1986 commented:
Gene, i've addressed your comments

peisun1986 commented:
Jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9219/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9220/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9221/
Test PASSed.

gpang commented:
Thanks for this improvement!

gpang commented:
LGTM

uronce-cc commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,23 @@ check_mount_mode() {
-    NoMount);;
+    NoMount)
+      if ! mount | grep ${ALLUXIO_RAM_FOLDER} > /dev/null; then
+        if [[ $( uname -a) == Darwin* ]]; then
```
NIT: `[[ $(uname -s) == Darwin ]]`

peisun1986 replied:
Done

uronce-cc commented on bin/alluxio-start.sh:
```diff
@@ -184,65 +201,74 @@ done
   worker)
-    check_mount_mode $2
-    start_worker $2
+    start_worker $1
```
`start_worker ${MOPT}`

peisun1986 replied:
done

uronce-cc commented on bin/alluxio-start.sh:
```diff
@@ -184,65 +201,74 @@ done
 
 # ensure log/data dirs
 ensure_dirs
 
```
I feel that replacing the following `$1` with `${FORMAT}` will improve readability.

peisun1986 replied:
good suggestion. Done.

uronce-cc commented:
@peisun1986 Thanks for the improvement! I left some comments.

peisun1986 commented:
@uronce-cc , I've addressed your comments. Thanks for the review

uronce-cc commented:
LGTM

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9234/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9236/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9235/
Test PASSed.

gpang commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,23 @@ check_mount_mode() {
+      fi
+      if [[ "${ALLUXIO_RAM_FOLDER}" =~ ^"/dev/shm"\/{0,1}$ ]]; then
+        echo "WARNING: Using tmpFS which is not guaranteed to be in memory."
+        echo "WARNING: Check vimstat for memory statistics (e.g. swapping)."
```
vimstat -> vmstat

peisun1115 replied:
DOne

gpang commented:
There is just a minor typo. Thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/9241/
Test PASSed.

gpang commented:
LGTM

yupeng9 commented:
Thanks for adding this feature! Shall we update the docs somewhere?

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -12,18 +12,19 @@ BIN=$(cd "$( dirname "$0" )"; pwd)
 USAGE="Usage: alluxio-start.sh [-hNw] WHAT [MOPT] [-f]
 Where WHAT is one of:
   all MOPT\t\tStart master and all workers.
```
is MOPT optional for `all`? if not, could we follow similar convention as for `local`?

peisun1115 replied:
Done.

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,23 @@ check_mount_mode() {
-    NoMount);;
+    NoMount)
+      if ! mount | grep "${ALLUXIO_RAM_FOLDER}" > /dev/null; then
+        if [[ $( uname -s) == Darwin ]]; then
```
nit: remove space between `$(` and `uname`

peisun1115 replied:
done

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -50,7 +51,23 @@ check_mount_mode() {
+      if ! mount | grep "${ALLUXIO_RAM_FOLDER}" > /dev/null; then
+        if [[ $( uname -s) == Darwin ]]; then
+          # Assuming Mac OS X
+          echo "ERROR: NoMount is not supported in Mac OS X."
```
s/in/on

peisun1115 replied:
done

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -199,50 +242,34 @@ ensure_dirs
     sleep 2
-    ${LAUNCHER} ${BIN}/alluxio-workers.sh ${BIN}/alluxio-start.sh worker $2
+
+    ${LAUNCHER} ${BIN}/alluxio-workers.sh ${BIN}/alluxio-start.sh worker ${MOPT}
```
quote paths

peisun1115 replied:
I am a little confused with the convection here.  I see lots of strings with variables unquoted in Alluxio bash scripts. Google style recommends quoting every string contains variable.  What is the convention that I should follow here?

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -199,50 +242,34 @@ ensure_dirs
   workers)
-    check_mount_mode $2
-    ${LAUNCHER} ${BIN}/alluxio-workers.sh ${BIN}/alluxio-start.sh worker $2 \
+    ${LAUNCHER} ${BIN}/alluxio-workers.sh ${BIN}/alluxio-start.sh worker ${MOPT} \
```
quote paths

peisun1115 replied:
done

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -184,12 +201,38 @@ done
+esac
+
+FORMAT=$1
+if [ ! -z ${FORMAT} ] && [ ${FORMAT} != "-f" ]; then
```
`[[ ]]` for consistency

peisun1115 replied:
done

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -12,18 +12,19 @@ BIN=$(cd "$( dirname "$0" )"; pwd)
 USAGE="Usage: alluxio-start.sh [-hNw] WHAT [MOPT] [-f]
```
replace `WHAT` with `\<action\>`

peisun1115 replied:
done

jsimsa commented on bin/alluxio-start.sh:
```diff
@@ -184,12 +201,38 @@ done
 shift $((OPTIND-1))
```
following google shell style guide, it would be great to put the script "body" into a main function: https://google.github.io/styleguide/shell.xml

peisun1115 replied:
done

jsimsa commented:
@peisun1115 thanks for the PR, left a number of comments related to shell style and readability

peisun1115 commented:
Jiri, apart from your comments, I also added some quotes in alluxio-start.sh. It seems that lots of the Alluxio scripts are not following Google style guide. I changed some in alluxio-start.sh. But not all of them as I don't know what is the general convention.

peisun1115 commented:
I have opened another PR 
https://github.com/Alluxio/alluxio/pull/3014/files
